specutils/spectra/spectral_axis.py

```diff
@@ -0,0 +1,98 @@
+import warnings
+
+import astropy.units as u
+from astropy.utils.decorators import lazyproperty
+import numpy as np
+
+from .spectral_coordinate import SpectralCoord
+
+
+__all__ = ['SpectralAxis']
+
+# We don't want to run doctests in the docstrings we inherit from Quantity
+__doctest_skip__ = ['SpectralAxis.*']
+
+class SpectralAxis(SpectralCoord):
+    """
+    Coordinate object representing spectral values corresponding to a specific
+    spectrum. Overloads SpectralCoord with additional information (currently
+    only bin edges).
+
+    Parameters
+    ----------
+    value : ndarray or `~astropy.units.Quantity` or `SpectralCoord`
+        Spectral axis data values.
+    unit : str or `~astropy.units.Unit`
+        Unit for the given data.
+    observer : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`, optional
+        The coordinate (position and velocity) of observer.
+    target : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`, optional
+        The coordinate (position and velocity) of observer.
+    radial_velocity : `~astropy.units.Quantity`, optional
+        The radial velocity of the target with respect to the observer.
+    redshift : float, optional
+        The redshift of the target with respect to the observer.
+    doppler_rest : `~astropy.units.Quantity`, optional
+        The rest value to use for velocity space transformations.
+    doppler_convention : str, optional
+        The convention to use when converting the spectral data to/from
+        velocity space.
+    bin_specificatoin: str, optional
+        Must be "edges" or "centers". Determines whether specified axis values
+        are interpreted as bin edges or bin centers. Defaults to "centers".
+    """
+
+    def __new__(cls, value, unit=None, observer=None, target=None,
+                 radial_velocity=None, redshift=None, doppler_rest=None,
+                 doppler_convention=None, bin_specification="centers",
+                 **kwargs):
+
+        # Convert to bin centers if bin edges were given, since SpectralCoord
+        # only accepts centers
+        if bin_specification == "edges":
+            bin_edges = value
+            value = SpectralAxis._centers_from_edges(value)
+
+        obj = super().__new__(cls, value, unit=unit, observer=observer,
+                              target=target, radial_velocity=radial_velocity,
+                              redshift=redshift, doppler_rest=doppler_rest,
+                              doppler_convention=doppler_convention, **kwargs)
+
+        if bin_specification == "edges":
+            obj.bin_edges = bin_edges
```

Comment:
_Technically_, this should conflict with the `bin_edges` property -- I'm not sure how this manifests, but ✨ just to be sure ✨ we may want to store this as `obj._bin_edges`.

```suggestion
            obj._bin_edges = bin_edges
```

Reply:
I wasn't sure if I grasped 100% what @lazyproperty does, but my hope was that if obj.bin_edges was already set then the `bin_edges` property would never be evaluated ( thus only evaluated if centers were specified). It seems to work as I expected in my testing.

Reply:
The `lazyproperty` in this case is just a lazy getter, meaning that the value of the property is computed only once, and later references simply use that same value, but otherwise it should be treated like a normal property getter. My concern in this case is that the property is _only_ a getter, and so setting it explicitly should technically not be allowed.

```python
In [7]: class Test():
   ...:     def __init__(self, *args, **kwargs):
   ...:         self._value = 0
   ...:
   ...:     @property
   ...:     def value(self):
   ...:         return self._value + 1
   ...:

In [8]:

In [8]: test = Test()

In [9]: test.value = 10
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-7baea496662b> in <module>
----> 1 test.value = 10

AttributeError: can't set attribute
```

It's a strange consequence of the `lazyproperty` decorator that it allows overriding the calculated value:

```python
In [10]: class Test():
    ...:     def __init__(self, *args, **kwargs):
    ...:         self._value = 0
    ...:
    ...:     @lazyproperty
    ...:     def value(self):
    ...:         return self._value + 1
    ...:

In [11]: test = Test()

In [12]: test.value
Out[12]: 1

In [13]: test.value = 10

In [14]: test.value
Out[14]: 10
```

Which it technically should not, since properties and values on instances should behave differently (likewise, it's a standard to not name a property the same as an instance attribute).

I think to maintain the "technically correct" part of this would mean that the `obj.bin_edges` attribute should be named something different.

Reply:
Makes sense, thanks. I'll update that today.